File: backend/service-ollama-router/app/client_tier_cache.py
"""Client tier cache — resolves max OpenRouter tier from MongoDB.

Router queries clients collection to find cloudModelPolicy.maxOpenRouterTier.
Cached with 5min TTL to avoid hitting DB on every route-decision.
"""

from __future__ import annotations

import logging
import time

from motor.motor_asyncio import AsyncIOMotorClient

from app.config import settings

logger = logging.getLogger("ollama-router.tier-cache")

# Cache: {client_id: (tier_string, timestamp)}
_cache: dict[str, tuple[str, float]] = {}
_mongo_client: AsyncIOMotorClient | None = None

# Default tier when client not found or DB unavailable
DEFAULT_TIER = "FREE"


async def _get_db():
    global _mongo_client
    if _mongo_client is None:
        _mongo_client = AsyncIOMotorClient(settings.mongodb_uri)
    return _mongo_client.get_default_database()


async def resolve_client_tier(client_id: str | None) -> str:
    """Resolve OpenRouter tier for a client. Returns tier string (NONE/FREE/PAID/PREMIUM).

    Uses in-memory cache with TTL. Falls back to DEFAULT_TIER on error.
    """
    if not client_id:
        return DEFAULT_TIER

    # Check cache
    cached = _cache.get(client_id)
    if cached:
        tier, ts = cached
        if time.monotonic() - ts < settings.client_tier_cache_ttl_s:
            return tier

    # Cache miss — query MongoDB
    try:
        db = await _get_db()
        from bson import ObjectId
        doc = await db["clients"].find_one(
            {"_id": ObjectId(client_id)},
            {"cloudModelPolicy.maxOpenRouterTier": 1},
        )
        if doc:
            policy = doc.get("cloudModelPolicy", {})
            tier = policy.get("maxOpenRouterTier", DEFAULT_TIER)
            if isinstance(tier, str):
                _cache[client_id] = (tier, time.monotonic())
                logger.info("Client tier resolved: %s → %s", client_id, tier)
                return tier

        # Client not found — default
        _cache[client_id] = (DEFAULT_TIER, time.monotonic())
        return DEFAULT_TIER

    except Exception as e:
        logger.warning("Failed to resolve tier for %s: %s — using default %s", client_id, e, DEFAULT_TIER)
        return DEFAULT_TIER


def invalidate_cache(client_id: str | None = None) -> None:
    """Invalidate cache for a client or all clients."""
    if client_id:
        _cache.pop(client_id, None)
    else:
        _cache.clear()


async def shutdown() -> None:
    """Close MongoDB connection."""
    global _mongo_client
    if _mongo_client:
        _mongo_client.close()
        _mongo_client = None

File: backend/service-ollama-router/src/main/kotlin/com/jervis/router/coord/ClientTierCache.kt
package com.jervis.router.coord

import com.mongodb.reactivestreams.client.MongoClient
import com.mongodb.reactivestreams.client.MongoClients
import kotlinx.coroutines.reactive.awaitFirstOrNull
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import mu.KotlinLogging
import org.bson.Document
import org.bson.types.ObjectId
import java.time.Instant

private val logger = KotlinLogging.logger {}

/**
 * Resolves a client's OpenRouter tier (NONE / FREE / PAID / PREMIUM) from
 * the `clients` collection. Cached in-memory with TTL.
 *
 * Mirrors Python `client_tier_cache.resolve_client_tier`. Falls back to
 * "FREE" on missing client / DB error so routing never crashes.
 */
class ClientTierCache(
    mongoUri: String,
    private val ttlS: Long,
) {
    private val mongoClient: MongoClient? = if (mongoUri.isBlank()) null else runCatching {
        MongoClients.create(mongoUri)
    }.onFailure { logger.warn { "Mongo client init failed: ${it.message}" } }.getOrNull()

    private val cache: MutableMap<String, Pair<String, Long>> = HashMap()
    private val mutex = Mutex()

    suspend fun resolve(clientId: String?): String {
        if (clientId.isNullOrBlank()) return DEFAULT_TIER
        mutex.withLock {
            val cached = cache[clientId]
            if (cached != null && (Instant.now().epochSecond - cached.second) < ttlS) {
                return cached.first
            }
        }

        val tier = lookup(clientId) ?: DEFAULT_TIER
        mutex.withLock { cache[clientId] = tier to Instant.now().epochSecond }
        logger.info { "Client tier resolved: $clientId → $tier" }
        return tier
    }

    fun invalidate(clientId: String?) {
        if (clientId == null) cache.clear() else cache.remove(clientId)
    }

    fun close() {
        mongoClient?.close()
    }

    private suspend fun lookup(clientId: String): String? {
        val client = mongoClient ?: return null
        return runCatching {
            val db = client.getDatabase("jervis")
            val coll = db.getCollection("clients")
            val objId = runCatching { ObjectId(clientId) }.getOrNull() ?: return@runCatching null
            val doc = coll.find(Document("_id", objId))
                .projection(Document("cloudModelPolicy.maxOpenRouterTier", 1))
                .first()
                .awaitFirstOrNull() ?: return@runCatching null
            val policy = doc.get("cloudModelPolicy") as? Document
            policy?.getString("maxOpenRouterTier")
        }.onFailure {
            logger.warn { "Failed to resolve tier for $clientId: ${it.message}" }
        }.getOrNull()
    }

    companion object {
        const val DEFAULT_TIER: String = "FREE"
    }
}
